### DISCUSSION TITLE
Promises returning null and empty array

### DISCUSSION THREAD

**Discussion by cmeffley** [nss-evening-cohort-14/evening-client-side#316] [Capstones!] [answered]
Chris Meffley

Both are close, but do not appear to help, that I can see
#304
#168

I’m working on my promise.all and my two promises that are being called in it are not returning what they should be:
`getRaceWorkouts` promise call returns and empty array.
`getSingleRace` promise call returns null

No error messages

https://stackoverflow.com/questions/53664314/promise-resolve-returns-empty-arrays-that-were-populated-using-array-push
https://stackoverflow.com/questions/47596195/promise-return-empty-array
Lots of Promise.all results in Google, but my issue is before that.

Tried:
-  `getRacesAthlete()` (or coach) first .then `axios.get` for single races
- `getSingleAthlete()` (or coach) first .then `axios.get` for single races
- Same as above for `getRaceWorkouts`
-Postman API calls are correct
In Firebase, I have added an indexOn for raceId to my workouts, even added firebaseKey to races.
raceId in workouts = firebaseKey in races
I believe it has something to do with checking which user is logged in. When console.warn athlete or coach, the correct object shows up. My check for which user is logged in still looks for the index . So do not know why it wouldn’t be looking for that users races/workouts.

Files:
https://github.com/cmeffley/front-end-capstone/blob/raceworkouts/src/helpers/data/raceData.js
https://github.com/cmeffley/front-end-capstone/blob/raceworkouts/src/helpers/data/workoutsData.js
https://github.com/cmeffley/front-end-capstone/blob/raceworkouts/src/helpers/data/raceWorkoutsData.js

https://github.com/cmeffley/front-end-capstone/blob/raceworkouts/src/App/index.js
https://github.com/cmeffley/front-end-capstone/blob/raceworkouts/src/helpers/data/coachData.js
https://github.com/cmeffley/front-end-capstone/blob/raceworkouts/src/helpers/data/athleteData.js

**Comment by drteresavasquez** [comment]
The null means Firebase isn't returning anything from that endpoint as in it may not exist and an empty array means there no data. Check your api calls again through postman to see what you get back.

**Reply by cmeffley** [reply]
Single Races - 
In raceData.js file: ```axios.get(`${dbUrl}/races/${firebaseKey}.json`)```
Get Postman call: ```databaseURL/races/-MbSHBgLZUwu6cqQChnW.json```
returns:
```
{
    "athleteUid": "Ve9GIoEnQ5YgrnGl4nafimyXMTE2",
    "coachUid": "yUuLS42kujRdkyU0U6RVTlIulaN2",
    "endDate": "October 17, 2021",
    "firebaseKey": "-MbSHBgLZUwu6cqQChnW",
    "raceDate": "October 17, 2021",
    "raceDistance": "5K",
    "raceLink": "https://sevenbridgesmarathon.com/5k",
    "raceName": "2 Bridges 5K",
    "startDate": "August 17, 2021"
}
```

Race Workouts -
In workoutsData.js file: ```axios.get(`${dbUrl}/workouts.json?orderBy="raceId"&equalTo="${raceId}"`)```
Get Postman call: ```databaseURL/workouts.json?orderBy="raceId"&equalTo="-MbSHBgLZUwu6cqQChnW"```
returns:
```
{
    "-MbSNHzPHzGGF5Hp1duu": {
        "actualWork": "I ran 1 mile and walked 1 mile",
        "athleteUid": "Ve9GIoEnQ5YgrnGl4nafimyXMTE2",
        "averagePace": 15,
        "coachUid": "yUuLS42kujRdkyU0U6RVTlIulaN2",
        "completed": true,
        "day": "Tuesday",
        "dueDay": "August 17, 2021",
        "firebaseKey": "-MbSNHzPHzGGF5Hp1duu",
        "plannedWork": "run-walk 2 miles at an easy pace",
        "raceId": "-MbSHBgLZUwu6cqQChnW",
        "startDay": "August 17, 2021",
        "totalMiles": 1,
        "workoutType": "Run"
    },
    "-MbSOAyb64S9GTzTyT9V": {
        "actualWork": "I did a quick workout today instead. I did QW 1",
        "athleteUid": "Ve9GIoEnQ5YgrnGl4nafimyXMTE2",
        "averagePace": 0,
        "coachUid": "yUuLS42kujRdkyU0U6RVTlIulaN2",
        "completed": true,
        "day": "Thursday",
        "dueDay": "September 9, 2021",
        "firebaseKey": "-MbSOAyb64S9GTzTyT9V",
        "plannedWork": "gym workout: do all the leg machines light weight for 21 reps",
        "raceId": "-MbSHBgLZUwu6cqQChnW",
        "startDay": "September 9, 2021",
        "totalMiles": 0,
        "workoutType": "Cross"
    }
}
```

Double checked my database url in my .env just in case, and that is correct. 
I'm sure I'm overlooking something

**Reply by drteresavasquez** [reply]
These are not the calls you mentioned in your post though. 

```javascript
const getRacesAthlete = (uid) => new Promise((resolve, reject) => {
  axios.get(`${dbUrl}/races.json?orderBy="athleteUid"&equalTo="${uid}"`)
    .then((response) => resolve(Object.values(response.data)))
    .catch((error) => reject(error));
});
```

What does this call give you? Make sure you are using the calls that are actually returning AND make sure you are actually passing the argument you defined in your function. You can use the debugger or log it. Tell me what you get for those.

**Reply by cmeffley** [reply]
There may be some confusion. The promise calls I am having issues with are `getRaceWorkouts` and `getSingleRace`. When I was troubleshooting, I included `getRacesAthlete` in the promise call of `getSingleRace` because I thought `getSingleRace` was not looking for the current users races.

To answer the question of what does `getRacesAthlete` give me, it returns the correct races.

**Comment by cmeffley** [comment, marked as answer]
I kept focusing on my axios call in my data file, but I needed to look at the broader picture of how everything was working...
1. My `history.push` was targeting the raceId in the workouts, but I needed to to be targeting the firebaseKey on the race (race firebaseKey = workouts raceId)
2. When calling my promise.all and setting state for my races and race workouts I was targeting the index of the race and race workouts. However, since I created a new object with keys of race and workout in my promise.all, I should have been targeting them. So when setting state with the responses, `setRace(response.race)` and `setRaceWorkout(response.workout)` instead of `(response[0])` or `(response[1])` which I thought was needed in order to target the races and workouts needed. But I was mistaken. Like I mentioned, I needed to focus on how I manipulated the data in the promise.all. Also, since I was doing a `.map` over my race workouts, that was a clue, that I shouldn't have targeted the indexes because I needed more than one object to appear.
